some functionality for G-sets

ThomasBreuer opened the pull request:
(to be dicussed)

See issue #108 for the background.
G-sets for matrix groups can be added when the notation from #298 becomes available (which vector objects shall be taken?).
Argument checks are missing.
This proposal is essentially about orbits of G-sets; stabilizers, transversals (e.g., the right transversal objects available in GAP), incomplete orbits (as in the GAP package Orb), etc. are not yet handled.

ThomasBreuer committed:
some functionality for G-sets

(to be dicussed)

fingolfin commented on src/Groups/gsets.jl:
```diff
@@ -23,18 +24,162 @@ export
 # - conjugacy classes of subgroups
 # - block system
-abstract type GSet{T} end
+abstract type GSet{TG,TE} end
```
Personally I prefer slightly more "self-explanatory" parameter names:
```suggestion
abstract type GSet{GrpT,ElmT} end
```

fingolfin commented on src/Groups/gsets.jl:
```diff
@@ -23,18 +24,162 @@ export
+    group::TG
+    action_function::Function
+    seeds::Set{TE}
+    AbstractAlgebra.@declare_other  # we want to store the orbits, the set of elements etc. if known
```
and as you wrote, later possibly also stabilizers, transversals, ...

fingolfin commented on test/Groups/gsets.jl:
```diff
@@ -0,0 +1,67 @@
+  @test ! isregular(Omega)
+  @test ! issemiregular(Omega)
+
+  Omega = gset(G, Set([[1, 2]]))  # action on ordered pairs
```
Do we really need to insist on the seeds being a `Set`? Any collection resp. itererable should do it, too, shouldn't it? With `collect` we can still turn that into a `Vector` and that then into a Set

fingolfin commented on test/Groups/gsets.jl:
```diff
@@ -0,0 +1,67 @@
+  Omega = gset(G, permuted, Set([omega]))
+  g = gens(G)[1]
+  x = Omega(omega)
+  @test x^g == Omega(permuted(omega, g))  # action via `^` is defined for `x`
```
It would be nice to also ahve a function to unwrap `x` again

thofma commented on src/Groups/gsets.jl:
```diff
@@ -23,18 +24,162 @@ export
 # - ...
 
+# G-set that is willing to write down complete orbits and elements set
+mutable struct GSet_By_Elements{TG,TE} <: GSet{TG,TE}
```
```suggestion
mutable struct GSetByElements{TG,TE} <: GSet{TG,TE}
```

ThomasBreuer committed:
general update

- changed `GSetByElements` to accept any iterable object as `seeds` (not just `Set`), only one type parameter seems to be meaningful now (the group type),
- introduced `gset_by_type` for dispatching on the element type of `seeds`
- added `unwrap`, `orbit` (via GAP, and a Julia variant), `permutation`, `action_homomorphism` (which needs a modification of GAP at runtime --to be dicussed)
- extended the tests

fingolfin commented on src/Groups/gsets.jl:
```diff
@@ -25,16 +30,307 @@ export
+      "local data; " *
+      "data:= JuliaData( hom ); " *
+      "return Julia.Oscar.permutation(data[1], Julia.Oscar.group_element(data[2], elm)).X;" *
+      "end );")
```
Perhaps use a triple quoted string? Those can contain newlines, so one can copy&paste between here and a terminal.

ThomasBreuer replied:
Yes, this would be a quick improvement.
In the long run, the question is where to put such "manipulations of GAP":
If `JuliaData` turns out to be useful in general, `JuliaInterface` is a natural place for its declaration.
The method used here depends on Oscar, thus `JuliaInterface` is not an option.
(Ideally, standard GAP could be changed/improved such that Oscar can just use it.)

ThomasBreuer committed:
added `representative_action` ...

... and `isconjugate`; cache `action_homomorphism`
